### PR TITLE
[SFSRAP-19] Allow use of pre-release packages and dependency versions.

### PR DESCRIPTION
**Pull request by HSAR**

### CODE REVIEW

**Comment by HSAR** [comment]
This test setup doesn't seem to correctly fail when the dependency cannot be found. Under investigation.

File: Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs [hunk 1]
```diff
@@ -17,22 +19,211 @@ public class NuGetPackageHandlerTests
...
+        private const string releaseVersion = "1.0.0";
+        private const string preReleaseVersion = "1.0.0-RC1";
+
         [TestMethod]
-        public void NuGet_TestDependencyResolutionFailure()
+        public void NuGet_TestPackageDownload_Release_Release()
         {
             // Arrange
             string testDir = TestUtils.CreateTestDirectory(this.TestContext);
+            string testDownloadDir = Path.Combine(testDir, "download");
```
Comment (bgavrilMS) [inline Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs]: Looks like a lot of the setup is duplicated between these tests. Consider refactoring into a separate method.

File: Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs [hunk 2]
```diff
@@ -17,22 +19,211 @@ public class NuGetPackageHandlerTests
...
+
+        #region Private Methods
+
+        private void AssertPackageDownloaded(string downloadDir, string packageName)
+        {
+            Assert.IsNotNull(Directory.GetDirectories(downloadDir).SingleOrDefault(d => d.Contains(packageName)),
+                "Expected a package to have been downloaded: " + packageName);
+        }
+
+        private void buildPackage(ManifestMetadata metadata, string destinationFile)
```
Comment (bgavrilMS) [inline Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs]: BuildPackage

File: Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs [hunk 3]
```diff
@@ -17,22 +19,211 @@ public class NuGetPackageHandlerTests
     {
         public TestContext TestContext { get; set; }
 
+        private const string testPackageName = "testPackage";
```
Comment (bgavrilMS) [inline Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs]: const fields should be UpperCase

File: Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs [hunk 4]
```diff
@@ -17,22 +19,211 @@ public class NuGetPackageHandlerTests
...
+                Description = "A description",
+                DependencySets = dependencies,
+            };
+            buildPackage(dependentMetadata, dependentPackageFile);
+
+            TestLogger logger = new TestLogger();
+            NuGetPackageHandler handler = new NuGetPackageHandler(logger);
+
+            // Act
+            // Attempt to download a package which is not released with a dependency that is released
```
Comment (bgavrilMS) [inline Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs]: Looks like package has a released version while the dependency has the pre-released version

File: Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs [hunk 5]
```diff
@@ -17,22 +19,211 @@ public class NuGetPackageHandlerTests
...
-            // No files should have been downloaded
-            Assert.IsTrue(Directory.GetFiles(testDir).Length == 0);
-            Assert.IsNull(package);
+            // Package should have been downloaded
+            AssertPackageDownloaded(testDownloadDir, dependentPackageName);
+            Assert.IsNotNull(package, "Expected a reference to the package to be returned");
         }
+
+        [TestMethod]
+        public void NuGet_TestPackageDownload_PreRelease_Release()
```
Comment (bgavrilMS) [inline Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs]: Should there be another test for the Release_PreRelease state?

File: Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs [hunk 6]
```diff
@@ -17,22 +19,211 @@ public class NuGetPackageHandlerTests
...
+            TestLogger logger = new TestLogger();
+            NuGetPackageHandler handler = new NuGetPackageHandler(logger);
+
+            // Act
+            // Attempt to download a package which is not released with a dependency that is not released
+            IPackage package = handler.FetchPackage(testDir, dependentPackageName, null, testDownloadDir);
+
+            // Assert
+            // Package should have been downloaded
+            AssertPackageDownloaded(testDownloadDir, dependentPackageName);
```
Comment (bgavrilMS) [inline Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs]: Is there any state where the package doesn't get downloaded, e.g. a package with a dependency that's not in NuGet or when the package itself is not present. Is this a good place for these kinds of tests or have these scenarios been tested elsewhere?

File: Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs (hunk 5 above)
Comment (HSAR) [inline Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs]: I'm not sure what happens when you try programmatically, but attempting to "nuget pack ..." in that scenario informs you that it is not allowed to produce a release package on a prerelease dependency.

File: Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs (hunk 4 above)
Comment (HSAR) [inline Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs]: Confusing naming convention, I think. The "dependent" package has a dependency on testPackage. In this case a prerelease package A depends on a released package B.

File: Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs (hunk 1 above)
Comment (duncanpMS) [inline Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs]: +1
It should be clear when reading a test which bits of data are important (the ids, version and dependency relationship) and which aren't (authors, description). At the moment the intent is lost in the noise.

File: Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs (hunk 6 above)
Comment (HSAR) [inline Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs]: Investigation is happening on those fronts. As far as I know, such scenarios are untested.

File: Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs [hunk 7]
```diff
@@ -17,22 +19,211 @@ public class NuGetPackageHandlerTests
...
-            IPackage package = handler.FetchPackage(AnalyzerPluginGenerator.NuGetPackageSource, "codeCracker", null, testDir);
+            // Attempt to download a package which is released with a dependency that is released
+            IPackage package = handler.FetchPackage(testDir, dependentPackageName, null, testDownloadDir);
 
             // Assert
-            // No files should have been downloaded
-            Assert.IsTrue(Directory.GetFiles(testDir).Length == 0);
-            Assert.IsNull(package);
+            // Package should have been downloaded
+            AssertPackageDownloaded(testDownloadDir, dependentPackageName);
```
Comment (duncanpMS) [inline Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs]: Check both packages have been downloaded.

File: Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs [hunk 8]
```diff
@@ -17,22 +19,211 @@ public class NuGetPackageHandlerTests
...
+            Assert.IsNotNull(package, "Expected a reference to the package to be returned");
```
Comment (duncanpMS) [inline Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs]: Repeated in every test -> factor out.
Checking the return value first (i.e. before checking the file system) seems more natural.

File: Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs (hunk 5 above)
Comment (bgavrilMS) [inline Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs]: Ok, good to know. Might want to write it down as a comment somewhere.

File: Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs (hunk 6 above)
Comment (bgavrilMS) [inline Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs]: Ok, don't worry about them for this fix.

File: Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs (hunk 7 above)
Comment (HSAR) [inline Tests/RoslynPluginGeneratorTests/NuGetPackageHandlerTests.cs]: In success cases only the downloaded package appears in the target directory.

**Pull request merged**